spock-core/src/main/java/org/spockframework/mock/runtime/ProxyBasedMockFactory.java

```diff
@@ -127,6 +142,42 @@ static Object createMock(final Class<?> type,
       ((ByteBuddyInterceptorAdapter.InterceptorAccess) proxy).$spock_set(interceptor);
       return proxy;
     }
+
+    private static boolean shouldLoadIntoCodegenPackage(Class<?> type) {
+      return isComingFromJDK(type) || isComingFromSignedJar(type) || isComingFromSealedPackage(type);
+    }
+
+    private static boolean isComingFromJDK(Class<?> type) {
+      // Try to read Implementation-Title entry from manifest which isn't present in every JDK JAR
+      return type.getPackage() != null && "Java Runtime Environment".equalsIgnoreCase(type.getPackage().getImplementationTitle())
+        || type.getName().startsWith("java.")
+        || type.getName().startsWith("javax.");
+    }
+
+    private static boolean isComingFromSealedPackage(Class<?> type) {
+      return type.getPackage() != null && type.getPackage().isSealed();
+    }
+
+    private static boolean isComingFromSignedJar(Class<?> type) {
+      return type.getSigners() != null;
+    }
+
+    @NotNull
+    private static ClassLoadingStrategy<ClassLoader> determineBestClassLoadingStrategy(Class<?> targetClass) throws Exception {
+      if (ClassInjector.UsingLookup.isAvailable()) {
+        Class<?> methodHandlesClass = Class.forName("java.lang.invoke.MethodHandles");
+        Class<?> lookupClass = Class.forName("java.lang.invoke.MethodHandles$Lookup");
+        Method lookupMethod = methodHandlesClass.getMethod("lookup");
+        Method privateLookupInMethod = methodHandlesClass.getMethod("privateLookupIn", Class.class, lookupClass);
+        Object lookup = lookupMethod.invoke(null);
+        Object privateLookup = privateLookupInMethod.invoke(null, targetClass, lookup);
+        return ClassLoadingStrategy.UsingLookup.of(privateLookup);
+      }
+      if (ClassInjector.UsingReflection.isAvailable()) {
+        return ClassLoadingStrategy.Default.INJECTION;
+      }
+      throw new IllegalStateException("No class loading strategy available");
```

Comment:
Shouldn't we fall back to `WRAPPER`

Reply:
In which cases would that get used?

Reply:
If a pre Java-9 JVM does not support Unsafe but I would not know any.